src/mbgl/tile/geometry_tile.cpp

```diff
@@ -114,6 +114,18 @@ std::vector<GeometryCollection> classifyRings(const GeometryCollection& rings) {
     return polygons;
 }
 
+void limitRings(GeometryCollection& polygon, uint32_t maxRings) {
+    if (polygon.size() > 1 + maxRings) {
+        std::nth_element(polygon.begin() + 1,
+                         polygon.begin() + 1 + maxRings,
+                         polygon.end(),
+                         [] (const auto& a, const auto& b) {
+                             return signedArea(a) > signedArea(b);
+                         });
+        polygon.resize(1 + maxRings);
```

Comment:
Looks good! The only minor thing is that `maxRings` name implies a limit to a total number of rings, including the outer ring, so you would need to remove a few `+1`s. Or rename to `maxHoles`.